sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,328 @@ def transform(self, X):
             X[coordinates] = values
 
         return X
+
+
+class KNNImputer(BaseEstimator, TransformerMixin):
+    """Imputation for completing missing values using k-Nearest Neighbors.
+
+    Each sample's missing values are imputed from up to "n_neighbors" nearest
+    neighbors found in the training set. Each missing feature is then
+    imputed as the average, either weighted or unweighted, of these neighbors
+    who have a value for it. Where all neighbors have that feature value
+    missing, the training set average for that feature is used for imputation.
+
+    Parameters
+    ----------
+    missing_values : integer or "NaN", optional (default = "NaN")
+        The placeholder for the missing values. All occurrences of
+        `missing_values` will be imputed. For missing values encoded as np.nan,
+        use the string value "NaN".
+
+    n_neighbors : int, optional (default = 5)
+        Maximum number of neighboring samples to use for imputation. When any
+        of the neighbors themselves have the feature value missing then the
+        remaining neighbors, if any, that have the feature value available are
+        used.
+
+    weights : str or callable, optional (default = "uniform")
+        Weight function used in prediction.  Possible values:
+
+        - 'uniform' : uniform weights.  All points in each neighborhood
+          are weighted equally.
+        - 'distance' : weight points by the inverse of their distance.
+          in this case, closer neighbors of a query point will have a
+          greater influence than neighbors which are further away.
+        - [callable] : a user-defined function which accepts an
+          array of distances, and returns an array of the same shape
+          containing the weights.
+
+    metric : str, optional (default = "masked_euclidean")
+        Distance metric for searching neighbors. Possible values:
+        - 'masked_euclidean'
+
+    row_max_missing : float, optional (default = 0.5)
```

Comment:
I'm not sure about using "row" and "column" here, where usually we use the terms "sample" and "feature". This could be `max_missing_features` and the other `max_missing_samples`, but I think that's a bit confusing. Could you have a think about whether there's a better way to name these?

Reply:
your suggestions make sense IMO. or if you like, perhaps a permuted version  features_max_missing/samples_max_missing? if not, i'll try to think of some more.